### ISSUE TITLE
[🐛 bug report] Differentiating the parameter `alpha` in `roughconductor` BSDF plugin got NaN after random iterations

### ISSUE THREAD

**Issue by binorchen** [mitsuba-renderer/mitsuba2#371]
## Summary
Hello, I am trying to differentiating with respect to the parameter **`alpha`** in **`roughconductor`** BSDF plugin. It looks ok in the beginning, but after *random* iterations, the value of **`alpha`** will become **nan**. It's really weird!

- Platform: Ubuntu 16.04
- CUDA: 11.0 with NVIDIA driver 450.66
- Compiler: Clang-11
- Python version: 3.6.3
- Mitsuba 2 version: 2.2.1
- Compiled variants:
    * `scalar_rgb`
    * `gpu_autodiff_rgb`

## Steps to reproduce

Please find the shared example from [HERE](https://drive.google.com/file/d/1kqtCpS2Fnys0uhOkMDH1nGvqWOWLQZz1/view?usp=sharing)
It is just a toy example to explain my issue. You can just run the `main.py`.

**Comment by Speierers** [comment]
Hi @binorchen ,

This is a common issue with the current system. Please make sure to sanatize the gradients you are computing and using in your gradient-descent step as those might contain `inf` or `nan` which will then propagate into the scene parameters you are optimizing.

We are working on making the AD computations more robust to `inf` and `nan` so this should be less of an issue in the upcoming version of Mitsuba 2 / Enoki.

**Comment by binorchen** [comment]
Thank you, @Speierers for your prompt reply!
I tried to use `torch.autograd.set_detect_anomaly(True)` duing the network training.
Please find the results as below:
```
  File "/home/vision/Projects/diff_render/src/train.py", line 92, in train
    img = render_torch(scene, params=params, unbiased=True, **params_torch)
  File "/home/vision/Softwares/mitsuba2/build/dist/python/mitsuba/python/autodiff.py", line 478, in render_torch
    *[num for elem in kwargs.items() for num in elem])
```

As you said the `autodiff.py` raises the problem.
I am wondering is there some possible quick solution to this? I would like to give it a try.

**Comment by Speierers** [comment]
You will need to figure out where those NaN originate from (forward rendering, gradient backpropagation, ...) and sanatize those arrays (e.g. replace the NaNs by `0.0`). This should already help

**Comment by binorchen** [comment]
Yes, I have used `torch.isnan` and `torch.isinf` to find and replace nan by zero_like number in the target and rendered images. But that's may not enough. I will trace deeper into this problem and let you guys know. Thanks a lot!

**Comment by Speierers** [comment]
Indeed this might not be enough as the NaNs could be generated by the AD backward AD computations. Good luck!

**Comment by binorchen** [comment]
@Speierers 
I finally find a way to partially solve this problem.
Use `torch.isnan` to check the target parameter. If `nan` is detected, 
- firstly exposed the parameters to PyTorch again `params_torch = params.torch()`, 
- then construct the optimizer again by `opt = torch.optim.Adam(params_torch.values(), lr = 0.001)`.

In my case:
```Python
if torch.isnan(params_torch['OBJMesh.bsdf.alpha.value']):
    params_torch = params.torch()
    opt = torch.optim.Adam(params_torch.values(), lr = 0.001)
```
Below shows the whole example.

```Python
import torch
import torchvision
from torch.autograd import Variable
import torchvision.transforms.functional as ttf

import mitsuba
mitsuba.set_variant('gpu_autodiff_rgb')
from mitsuba.core import Thread, Vector3f, Float
from mitsuba.core.xml import load_file
from mitsuba.python.util import traverse
from mitsuba.python.autodiff import render_torch, write_bitmap


def train():
    # Load the scene
    Thread.thread().file_resolver().append('./')
    scene = load_file('blob.xml')
    # Specify which parameters needs to be optimized
    params = traverse(scene)
    print(params)
    params.keep(['OBJMesh.bsdf.alpha.value'])

    img_ref = render_torch(scene)
    crop_size = scene.sensors()[0].film().crop_size()
    write_bitmap('./ref_out.png', img_ref, crop_size)

    params['OBJMesh.bsdf.alpha.value'] = 0.01
    params.update()
    params_torch = params.torch()
    opt = torch.optim.Adam(params_torch.values(), lr = 0.001)

    mse = torch.nn.MSELoss()  # Use MSE loss

    for epoch in range(1000):

        if torch.isnan(params_torch['OBJMesh.bsdf.alpha.value']):
            params_torch = params.torch()
            opt = torch.optim.Adam(params_torch.values(), lr = 0.001)

        opt.zero_grad()
        # Do the differentiable rendering
        img = render_torch(scene, params=params, unbiased=True, **params_torch)

        if epoch % 10 == 0:
            write_bitmap('./output/out_%03i.png' % epoch, img, crop_size)

        loss_img = mse(img, img_ref)
        loss_img.backward()
        opt.step()

        print('Iteration %03i: img_error=%g' % (epoch, loss_img.cpu().detach()))
        print(params_torch['OBJMesh.bsdf.alpha.value'])


if __name__ == "__main__":
    train()
```

I think this method can adapt to similar problems related to `NaN`.
Hope this can help someone. 
Cheers!

**Comment by HsiangYangChu** [comment]
Hi @binorchen,
Thanks for your code first! I have used your code and got better results! But I found that `inf` appeared, but the results are still ok. So, my question is how to deal with the ‘inf’?
Looking forward to your insights!
